File: starter_kits/Ruby/hlt/map.rb
require 'direction'
require 'position'
require 'cell'


# The game map.
# Can be indexed by a position, or by a contained entity.
# Coordinates start at 0. Coordinates are normalized for you.
class Map

  def initialize(cells, width, height)
    @width = width
    @height = height
    @cells = cells
  end

  # Getter for position object or entity objects within the game map
  # :param location: the position or entity to access in this map
  # :return: the contents housing that cell or entity
  def [](location)
    if location.is_a? Position
      location = normalize(location)
      return @cells[location.y][location.x]
    elsif location.is_a? Entity
      return @cells[location.position.y][location.position.x]
    end
    return nil
  end

  # Compute the Manhattan distance between two locations.
  #   Accounts for wrap-around.
  # :param source: The source from where to calculate
  # :param target: The target to where calculate
  # :return: The distance between these items
  def calculate_distance(source, target)
    source = normalize(source)
    target = normalize(target)
    resulting_position = (source - target).abs
    return [resulting_position.x, @width - resulting_position.x].min +
      [resulting_position.y, @height - resulting_position.y].min
  end

  # Normalize the position within the bounds of the toroidal map.
  #   i.e.: Takes a point which may or may not be within width and
  #   height bounds, and places it within those bounds considering
  #   wraparound.
  # :param position: A position object.
  # :return: A normalized position object fitting within the bounds of the map
  def normalize(position)
    Position.new(position.x % @width, position.y % @height)
  end

  # Return the Direction(s) to move closer to the target point, or empty if the
  #   points are the same.
  #   This move mechanic does not account for collisions. The multiple
  #   directions are if both directional movements are viable.
  # :param source: The starting position
  # :param destination: The destination towards which you wish to move your object.
  # :return: A list of valid (closest) Directions towards your target.
  def get_unsafe_moves(source, destination)
    source = normalize(source)
    destination = normalize(destination)
    possible_moves = []
    distance = (destination - source).abs
    y_cardinality, x_cardinality = self.class.get_target_direction(source, destination)

    if distance.x != 0
      possible_moves << (distance.x < (@width / 2) ? x_cardinality :
                  Direction.invert(x_cardinality))
    end
    if distance.y != 0
      possible_moves << (distance.y < (@height / 2) ? y_cardinality :
                  Direction.invert(y_cardinality))
    end
    return possible_moves
  end

  # Returns a singular safe move towards the destination.
  # :param ship: The ship to move.
  # :param destination: Ending position
  # :return: A direction.
  def naive_navigate(ship, destination)
    # No need to normalize destination, since get_unsafe_moves does that
    for direction in get_unsafe_moves(ship.position, destination)
      target_pos = ship.position.directional_offset(direction)
      if !self[target_pos].is_occupied?
        self[target_pos].mark_unsafe(ship)
        return direction
      end
    end

    return Direction::STILL
  end

  # Updates this map object from the input given by the game engine
  # :return: nothing
  def update(game)
    # Mark cells as safe for navigation (will re-mark unsafe cells later)
    for y in 0...@height
      for x in 0...@width
        self[Position.new(x, y)].ship = nil
      end
    end

    game.read_ints_from_input.first.times do
      cell_x, cell_y, cell_energy = game.read_ints_from_input
      self[Position.new(cell_x, cell_y)].halite_amount = cell_energy
    end
  end

  # Creates a map object from the input given by the game engine
  # :return: The map object
  def self.generate(game)
    map_width, map_height = game.read_ints_from_input
    game_map = []
    for y_pos in 0...map_height
      game_map[y_pos] = []
      cells = game.read_ints_from_input
      for x_pos in 0...map_width
        game_map[y_pos][x_pos] = Cell.new(Position.new(x_pos, y_pos), cells[x_pos])
      end
    end
    return Map.new(game_map, map_width, map_height)
  end

  # Returns where in the cardinality spectrum the target is from source.
  # e.g.: North, East; South, West; etc.
  # NOTE: Ignores toroid
  # :param source: The source position
  # :param target: The target position
  # :return: An array containing the target Direction(s). Either item (or both)
  #   could be nil if within same coords
  def self.get_target_direction(source, target)
    y_cardinality = target.y > source.y ? Direction::SOUTH : (target.y < source.y ? Direction::NORTH : nil)
    x_cardinality = target.x > source.x ? Direction::EAST  : (target.x < source.x ? Direction::WEST : nil)
    return [y_cardinality, x_cardinality]
  end
end

File: starter_kits/Swift/Halite-III/Halite-III/hlt/Map.swift
//
//  Map.swift
//  Halite-III
//
//  Created by Chris Downie on 10/24/18.
//  Copyright © 2018 Chris Downie. All rights reserved.
//

import Foundation

/// Gameplay takes place on a wrapping rectangular grid 32x32, 40x40, 48x48, 56x56, or 64x64 in dimension. The map
/// edges wrap to their opposite edge and create a torus shape. The game map can be indexed by a position or by a
/// contained entity (ship, shipyard, or dropoff). The game map has width and height as member variables.
struct Map {
    let width: Int
    let height: Int
    
    private var cells: [[Cell]]
    var haliteGrid: [[Int]] {
        return cells.map { $0.map { $0.haliteAmount } }
    }
    
    init(width: Int, height: Int, initialHalite: [[Int]]? = nil) {
        self.width = width
        self.height = height
        self.cells = (0 ..< height).map { row in
            (0 ..< width).map { column in
                Cell(position: Position(x: column, y: row),
                     haliteAmount: initialHalite?[row][column] ?? 0,
                     ship: nil,
                     structure: nil)
            }
        }
    }
    
    /// A map cell is an object representation of a cell on the game map.
    class Cell: Placeable {
        let position: Position
        let haliteAmount: Int
        var ship: Ship?
        var structure: Structure?
        
        init(position: Position, haliteAmount: Int, ship: Ship?, structure: Structure?) {
            self.position = position
            self.haliteAmount = haliteAmount
            self.ship = ship
            self.structure = structure
        }
        
        /// Returns True if the cell is empty.
        var isEmpty: Bool {
            return ship == nil && structure == nil
        }
        
        /// Returns True if there is a ship on this cell.
        var isOccupied: Bool {
            return ship != nil
        }
        
        /// Returns True if there is a structure (a dropoff or shipyard) on this cell.
        var hasStructure: Bool {
            return structure != nil
        }
        
        /// Returns the type of structure on this cell, or nil if there is no structure.
        var structureType: StructureType? {
            return structure?.type
        }
        
        /// Used to mark the cell under this ship as unsafe (occupied) for collision avoidance.
        ///
        /// This marking resets every turn and is used by naive_navigate to avoid collisions.
        ///
        /// - Parameter ship: The ship on this cell.
        func markUnsafe(ship: Ship) {
            self.ship = ship
        }
    }
    
    /// A method that computes the Manhattan distance between two locations, and accounts for the toroidal wraparound.
    ///
    /// - Parameters:
    ///   - source: The position to measure from
    ///   - target: The position to measure to
    /// - Returns: The distance between the source and the target
    func calculateDistance(source: Position, target: Position) -> Int {
        let from = normalize(position: source)
        let to = normalize(position: target)
        let horizontalDistance = abs(from.x - to.x)
        let verticalDistance = abs(from.y - to.y)
        
        return min(horizontalDistance, width - horizontalDistance)
            + min(verticalDistance, height - verticalDistance)
    }
    
    /// A method that normalizes a position within the bounds of the toroidal map.
    ///
    /// Useful for handling the wraparound modulus arithmetic on x and y. For example, if a ship at (x = 31, y = 4)
    /// moves to the east on a 32x32 map, the normalized position would be (x = 0, y = 4), rather than the off-the-map
    /// position of (x = 32, y = 4).
    ///
    /// - Parameter position: The position to normalize
    /// - Returns: The normalized position.
    func normalize(position: Position) -> Position {
        var x = position.x
        var y = position.y
        
        // In Swift, % is a remainder operator, not modulus. So we have to get a positive number to get the wrap
        // around to work correctly.
        while x < 0 {
            x += width
        }
        while y < 0 {
            y += height
        }
        
        return Position(x: x % width, y: y % height)
    }
    
    /// A method that returns a list of direction(s) to move closer to a target disregarding collision possibilities.
    /// Returns an empty list if the source and destination are the same.
    ///
    /// - Parameters:
    ///   - source: The position to move from
    ///   - destination: The position to move to
    /// - Returns: List of closest directions toward the given target.
    func getUnsafeMoves(source: Position, destination: Position) -> [Direction] {
        let from = normalize(position: source)
        let to = normalize(position: destination)
        let xDistance = to.x - from.x
        let yDistance = to.y - from.y
        let xCardinality: Direction = xDistance > 0 ? .east : .west
        let yCardinality: Direction = yDistance > 0 ? .south : .north
        
        var possibleMoves = [Direction]()
        
        if xDistance != 0 {
            if Double(abs(xDistance)) < Double(width) / 2.0 {
                possibleMoves.append(xCardinality)
            } else {
                possibleMoves.append(Direction.invert(xCardinality))
            }
        }
        
        if yDistance != 0 {
            if Double(abs(yDistance)) < Double(width) / 2.0 {
                possibleMoves.append(yCardinality)
            } else {
                possibleMoves.append(Direction.invert(yCardinality))
            }
        }
        
        return possibleMoves
    }
    
    /// A method that returns a direction to move closer to a target without colliding with other entities.
    ///
    /// Returns a direction of “still” if no such move exists.
    ///
    /// - Parameters:
    ///   - ship: The ship attempting to navigate
    ///   - destination: The ship's desired position.
    /// - Returns: A single valid direction toward a given target.
    func naiveNavigate(ship: Ship, destination: Position) -> Direction {
        // No need to normalize destination, since getUnsafeMoves does that
        let direction = getUnsafeMoves(source: ship.position, destination: destination).first { direction in
            let targetPosition = ship.position.directionalOffset(direction)
            return !self[targetPosition].isOccupied
        }
        
        if let direction = direction {
            let cell = self[ship.position.directionalOffset(direction)]
            cell.markUnsafe(ship: ship)
        }
        
        return direction ?? .still
    }
    
    /// Index the game map and find a particular map cell with game_map[position].
    ///
    /// - Parameter position: The position to return a map cell
    subscript(_ position: Position) -> Map.Cell {
        get {
            let normalized = normalize(position: position)
            return cells[normalized.y][normalized.x]
        }
    }
    
    // Mark: - Convenience methods
    subscript(_ placeable: Placeable) -> Map.Cell {
        get {
            return self[placeable.position]
        }
    }
    
    func unsafeMoves(from: Position, to: Position) -> [Direction] {
        return getUnsafeMoves(source: from, destination: to)
    }
}

extension Map {
    /// Computes the Manhattan distance between two locations, and accounts for the toroidal wraparound.
    ///
    /// This is a convenience method on calculateDistance(source:target) the fits the Swift language better
    /// - Parameters:
    ///   - from: The position to measure from
    ///   - to: The position to measure to
    /// - Returns: The manhattan distance between those two positions
    func distance(from: Position, to: Position) -> Int {
        return calculateDistance(source: from, target: to)
    }
}
